ref(alert): Add trailingItems prop to <Alert />

vuluongj20 opened the pull request:
A continuation of https://github.com/getsentry/sentry/pull/32779.

Add a `trailingItems` prop to `<Alert />`. This is useful for adding elements to the right side of alerts (like the "Upgrade now" button in the screenshot below) while making sure that they are properly aligned with the main text.

<img width="633" alt="pluginDeprecation--after" src="https://user-images.githubusercontent.com/44172267/159073812-603a9a73-4bb1-42c8-8f4d-bf81c23bc633.png">

Before-after screenshots available in the comments.

vuluongj20 committed:
ref(alerts): Update <Alert /> UI

vuluongj20 committed:
ref(alerts): Remove usages of `icon` prop in <Alert />

vuluongj20 committed:
ref(permission-alert): Show alert icon

vuluongj20 committed:
ref(alert): Add `trailingItems` prop to <Alert />

vuluongj20 committed:
fix(alert): Fix trailingItems' type

vuluongj20 committed:
ref(alert): Reduce padding-right is trailingItems is present

vuluongj20 committed:
ref(alerts): Use trailingItems prop in alerts

vuluongj20 committed:
ref(widget-viewer-modal): Use showIcon prop

vuluongj20 committed:
fix(tests): Fix processingIssueHint test

vuluongj20 committed:
ref(storybook): Update Alert page

vuluongj20 commented on static/app/components/forms/formField/index.tsx:
```diff
@@ -406,24 +404,29 @@ class FormField extends React.Component<FormFieldProps> {
-                        </Button>
-                      </ButtonBar>
-                    </MessageAndActions>
+                  <PanelAlert
+                    type={saveMessageAlertType}
+                    trailingItems={[
+                      <Button
+                        onClick={this.handleCancelField}
+                        size="xsmall"
+                        key="cancel-button"
+                      >
+                        {t('Cancel')}
+                      </Button>,
+                      <Button
+                        priority="primary"
+                        size="xsmall"
+                        onClick={this.handleSaveField}
+                        key="save-button"
+                      >
+                        {t('Save')}
+                      </Button>,
+                    ]}
+                  >
+                    {typeof saveMessage === 'function'
+                      ? saveMessage({...props, value})
+                      : saveMessage}
```
<img width="735" alt="formField--before" src="https://user-images.githubusercontent.com/44172267/159083241-0a8756c4-0c77-49b5-b4d9-64f67d181d69.png">
<img width="735" alt="formField--after" src="https://user-images.githubusercontent.com/44172267/159083249-77c6f289-bba6-4ffd-984c-9a6140a51cde.png">

vuluongj20 commented on static/app/components/globalSdkUpdateAlert.tsx:
```diff
@@ -89,45 +87,37 @@ function InnerGlobalSdkUpdateAlert(
-          </Button>
-        </Actions>
-      </Content>
+    <Alert
+      type="info"
+      showIcon
+      trailingItems={[
+        <Button
+          priority="link"
+          size="zero"
+          title={t('Dismiss for the next two weeks')}
+          onClick={handleSnoozePrompt}
+          key="dismiss-button"
+        >
+          {t('Remind me later')}
+        </Button>,
+        <span key="divider">|</span>,
+        <Button
+          priority="link"
+          size="zero"
+          onClick={handleReviewUpdatesClick}
+          key="review-button"
+        >
+          {t('Review updates')}
+        </Button>,
+      ]}
+    >
+      {t(
+        `You have outdated SDKs in your projects. Update them for important fixes and features.`
+      )}
```
<img width="958" alt="globalSDKUpdate--before" src="https://user-images.githubusercontent.com/44172267/159083276-4e537b99-8724-4e36-adc2-9fe2a8c34bf1.png">
<img width="958" alt="globalSDKUpdate--after" src="https://user-images.githubusercontent.com/44172267/159083282-52fae0b2-5fd6-4379-aaad-0e601567103c.png">

vuluongj20 commented on static/app/components/loadingError.tsx:
```diff
@@ -31,16 +29,18 @@ class LoadingError extends React.Component<Props> {
-          {onRetry && (
+      <StyledAlert
+        type="error"
+        showIcon
+        trailingItems={
+          onRetry && (
             <Button onClick={onRetry} type="button" priority="default" size="small">
               {t('Retry')}
             </Button>
-          )}
-        </Content>
+          )
+        }
+      >
+        <div data-test-id="loading-error-message">{message}</div>
```
<img width="759" alt="loadingError--before" src="https://user-images.githubusercontent.com/44172267/159083312-fba396a9-90f3-4cb6-8024-31d58fbfc4b2.png">
<img width="759" alt="loadingError--after" src="https://user-images.githubusercontent.com/44172267/159083315-9e401151-08fe-48b0-b8fc-52891232e7d2.png">

vuluongj20 commented on static/app/components/organizations/pageFilters/desyncedFiltersAlert.tsx:
```diff
@@ -31,39 +30,37 @@ export default function DesyncedFilterAlert({router}: Props) {
-          />
-        </ButtonBar>
-      </AlertWrapper>
+    <Alert
+      type="info"
+      showIcon
+      system
+      trailingItems={[
+        <RevertButton
+          key="revert-button"
+          priority="link"
+          size="zero"
+          onClick={onRevertClick}
+          borderless
+        >
+          {t('Revert')}
+        </RevertButton>,
+        <Button
+          key="close-button"
+          priority="link"
+          size="zero"
+          icon={<IconClose color="purple300" />}
+          aria-label={t('Close Alert')}
+          onClick={() => setHideAlert(true)}
+        />,
+      ]}
+    >
+      {t(
+        "You're viewing a shared link. Certain queries and filters have been automatically filled from URL parameters."
+      )}
```
<img width="972" alt="pageFilterSyncAlert--before" src="https://user-images.githubusercontent.com/44172267/159083369-3c0b4283-0053-4750-b72c-dc86b7e5adca.png">
<img width="972" alt="pageFilterSyncAlert--after" src="https://user-images.githubusercontent.com/44172267/159083366-043471c3-f5f4-4c3e-8b3b-27eac3c1babe.png">

vuluongj20 commented on static/app/components/stream/processingIssueHint.tsx:
```diff
@@ -69,19 +69,18 @@ function ProcessingIssueHint({orgId, projectId, issue, showProject}: Props) {
-          </div>
-        )}
-      </Wrapper>
+    <StyledAlert
+      type={alertType}
+      showIcon
+      trailingItems={
+        showButton && (
+          <StyledButton size="xsmall" to={link}>
+            {t('Show details')}
+          </StyledButton>
+        )
+      }
+    >
+      {project} <strong>{text}</strong> {lastEvent}
```
<img width="760" alt="processingIssueHint--before" src="https://user-images.githubusercontent.com/44172267/159083416-b247f7e4-1ca4-4edc-81c0-1c48f041aea0.png">
--
<img width="760" alt="processingIssueHint--after" src="https://user-images.githubusercontent.com/44172267/159083424-64078844-c31a-4959-b3e7-132d2adf378c.png">

vuluongj20 commented on static/app/components/suggestProjectCTA.tsx:
```diff
@@ -190,16 +189,14 @@ class SuggestProjectCTA extends Component<Props, State> {
-          </span>
-          <StyledIconClose onClick={this.handleCTAClose} />
-        </Content>
+      <Alert
+        type="info"
+        trailingItems={<StyledIconClose onClick={this.handleCTAClose} />}
+      >
+        {tct(
+          'We have a sneaking suspicion you have a mobile app that doesn’t use Sentry. [link:Start Monitoring]',
+          {link: <a onClick={this.openModal} />}
+        )}
```
<img width="929" alt="suggestProjectCTA--before" src="https://user-images.githubusercontent.com/44172267/159083477-010eb885-5bb4-49f5-bcb9-4306a30d9720.png">
<img width="929" alt="suggestProjectCTA--after" src="https://user-images.githubusercontent.com/44172267/159083484-ea834cf6-b4ea-4f4c-a7c7-489e42812da7.png">

vuluongj20 commented on static/app/views/alerts/rules/details/relatedIssuesNotAvailable.tsx:
```diff
@@ -21,18 +19,20 @@ export const RELATED_ISSUES_BOOLEAN_QUERY_ERROR =
-      </div>
+  <StyledAlert
+    type="info"
+    showIcon
+    trailingItems={
       <Feature features={['discover-basic']}>
-        <Button type="button" priority="default" size="small" to={buttonTo}>
+        <Button type="button" priority="default" size="xsmall" to={buttonTo}>
           {buttonText}
         </Button>
       </Feature>
-    </Content>
+    }
+  >
+    <div data-test-id="loading-error-message">
+      Related Issues unavailable for this alert.
+    </div>
```
<img width="955" alt="relateIssuesNotAvailable--before" src="https://user-images.githubusercontent.com/44172267/159083514-d8375d0e-fbd5-4525-b587-94e0ff80ae40.png">
<img width="955" alt="relateIssuesNotAvailable--after" src="https://user-images.githubusercontent.com/44172267/159083516-93933d44-712e-4be7-b2fa-d9acb3abae02.png">

vuluongj20 commented on static/app/views/app/alertMessage.tsx:
```diff
@@ -21,17 +21,22 @@ const AlertMessage = ({alert, system}: Props) => {
-        size="zero"
-        borderless
-      />
+    <StyledAlert
+      type={type}
+      showIcon
+      system={system}
+      opaque={opaque}
+      trailingItems={
+        <StyledCloseButton
+          icon={<IconClose size="sm" />}
+          aria-label={t('Close')}
+          onClick={alert.onClose ?? handleClose}
+          size="zero"
+          borderless
+        />
+      }
+    >
+      {url ? <ExternalLink href={url}>{message}</ExternalLink> : message}
```
<img width="974" alt="appAlertMessage--before" src="https://user-images.githubusercontent.com/44172267/159083546-14a603ff-665f-4bd3-9034-7c2d9892240f.png">
<img width="974" alt="appAlertMessage--after" src="https://user-images.githubusercontent.com/44172267/159083557-2680ec8a-4e8a-46b4-890c-2cd458e92806.png">

vuluongj20 commented on static/app/views/organizationGroupDetails/quickTrace/issueQuickTrace.tsx:
```diff
@@ -133,7 +127,16 @@ class IssueQuickTrace extends Component<Props, State> {
+                {t('Read the docs to understand why.')}
+              </ExternalLink>
+            ),
+          })}
```
<img width="860" alt="issueQuickTrace--before" src="https://user-images.githubusercontent.com/44172267/159083610-5f6f466d-3bb4-48e2-a0fa-e2bf08329e14.png">
<img width="860" alt="issueQuickTrace--after" src="https://user-images.githubusercontent.com/44172267/159083613-a6c5379e-cffb-4500-b43d-84eafd721f26.png">

vuluongj20 commented on static/app/views/organizationIntegrations/integrationRow.tsx:
```diff
@@ -119,21 +119,26 @@ const IntegrationRow = (props: Props) => {
-            >
-              {resolveText || t('Resolve Now')}
-            </ResolveNowButton>
+          <Alert
+            type="warning"
+            showIcon
+            trailingItems={
+              <ResolveNowButton
+                href={`${baseUrl}?tab=configurations&referrer=directory_resolve_now`}
+                size="xsmall"
+                onClick={() =>
+                  trackIntegrationAnalytics('integrations.resolve_now_clicked', {
+                    integration_type: convertIntegrationTypeToSnakeCase(type),
+                    integration: slug,
+                    organization,
+                  })
+                }
+              >
+                {resolveText || t('Resolve Now')}
+              </ResolveNowButton>
+            }
+          >
+            {alertText}
```
<img width="610" alt="integrationRow--before" src="https://user-images.githubusercontent.com/44172267/159083648-899be922-8cfd-48b6-b68f-b376558dc31c.png">
<img width="610" alt="integrationRow--after" src="https://user-images.githubusercontent.com/44172267/159083653-1a910847-13ec-4684-bc4c-10a79031abf6.png">

vuluongj20 commented on static/app/views/organizationIntegrations/pluginDeprecationAlert.tsx:
```diff
@@ -29,21 +29,26 @@ class PluginDeprecationAlert extends Component<Props, State> {
-          >
-            {t('Upgrade Now')}
-          </UpgradeNowButton>
+        <Alert
+          type="warning"
+          showIcon
+          trailingItems={
+            <UpgradeNowButton
+              href={`${upgradeUrl}${queryParams}`}
+              size="xsmall"
+              onClick={() =>
+                trackIntegrationAnalytics('integrations.resolve_now_clicked', {
+                  integration_type: 'plugin',
+                  integration: plugin.slug,
+                  organization,
+                })
+              }
+            >
+              {t('Upgrade Now')}
+            </UpgradeNowButton>
+          }
+        >
+          {`This integration is being deprecated on ${plugin.deprecationDate}. Please upgrade to avoid any disruption.`}
```
<img width="633" alt="pluginDeprecation--before" src="https://user-images.githubusercontent.com/44172267/159083675-87986359-ac76-4c0d-a0af-1a4c2a302266.png">
<img width="633" alt="pluginDeprecation--after" src="https://user-images.githubusercontent.com/44172267/159083678-90189ed1-6674-4960-bcf0-56dfdbdaf134.png">

evanpurkhiser commented on static/app/components/createAlertButton.tsx:
```diff
@@ -441,9 +446,7 @@ const StyledCode = styled('code')`
 
 const StyledCloseButton = styled(Button)`
```
Kind of looks like we can just kill this component?

vuluongj20 replied:
oh true! done in https://github.com/getsentry/sentry/pull/32781/commits/d0022f9a26621d7d4709ef742bb28128d8fd2f31

evanpurkhiser commented on static/app/components/globalSdkUpdateAlert.tsx:
```diff
@@ -89,45 +87,37 @@ function InnerGlobalSdkUpdateAlert(
+    <Alert
+      type="info"
+      showIcon
+      trailingItems={[
+        <Button
+          priority="link"
+          size="zero"
+          title={t('Dismiss for the next two weeks')}
+          onClick={handleSnoozePrompt}
+          key="dismiss-button"
+        >
+          {t('Remind me later')}
+        </Button>,
+        <span key="divider">|</span>,
+        <Button
+          priority="link"
+          size="zero"
+          onClick={handleReviewUpdatesClick}
+          key="review-button"
+        >
+          {t('Review updates')}
+        </Button>,
+      ]}
+    >
```
Use `<Fragment>` instead of giving things keys

evanpurkhiser commented on static/app/components/forms/formField/index.tsx:
```diff
@@ -406,24 +404,29 @@ class FormField extends React.Component<FormFieldProps> {
-                    </MessageAndActions>
+                  <PanelAlert
+                    type={saveMessageAlertType}
+                    trailingItems={[
```
Same comment about <Fragment /> here

vuluongj20 committed:
fix(alerts): Restore ref-[type] class name

vuluongj20 committed:
ref(alerts): Use React Fragment in trailingItems prop

getsantry[bot] committed:
style(lint): Auto commit lint changes

vuluongj20 committed:
ref(create-alert-button): Remove StyledCloseButton

The styles are no longer necessary given the automatic alignment from Alert's trailingItems prop.

priscilawebdev commented on static/app/components/forms/formField/index.tsx:
```diff
@@ -406,24 +404,31 @@ class FormField extends React.Component<FormFieldProps> {
-                          type="button"
+                          size="xsmall"
                           onClick={this.handleSaveField}
+                          key="save-button"
```
do you need these keys?

vuluongj20 replied:
oh true, no need for these. they're from before when I put in an array instead of a `Fragment`.

priscilawebdev commented on static/app/components/loadingError.tsx:
```diff
@@ -31,16 +29,18 @@ class LoadingError extends React.Component<Props> {
+          )
+        }
+      >
+        <div data-test-id="loading-error-message">{message}</div>
```
instead of using a test-id, what do you think we only check for the text? `screen.getByText()`

vuluongj20 replied:
ahh good point! https://github.com/getsentry/sentry/pull/32781/commits/264dcfcfbdaab13762cb9fa9e1cba89d78b75fa9

priscilawebdev commented on static/app/components/suggestProjectCTA.tsx:
```diff
@@ -190,16 +189,14 @@ class SuggestProjectCTA extends Component<Props, State> {
+      >
+        {tct(
+          'We have a sneaking suspicion you have a mobile app that doesn’t use Sentry. [link:Start Monitoring]',
+          {link: <a onClick={this.openModal} />}
```
should we use a Button with a priority link here?

vuluongj20 replied:
hmm here I think we can use a normal button and move it to `trailingItems`, https://github.com/getsentry/sentry/pull/32781/commits/1c957318f6ad4306aca8eabd8259e0b95f10c1c0:
<img width="931" alt="Screen Shot 2022-03-24 at 10 37 05 AM" src="https://user-images.githubusercontent.com/44172267/159976861-a4ba499d-cb01-4332-8540-8460a00bcb9e.png">

vuluongj20 committed:
style(alert): Remove unnecessary keys

No longer needed because of 07d32831ee6ce18faa028c699f03e8afbc6a70ab

vuluongj20 committed:
ref(loading-error): Use alert text instead of data-test-id

vuluongj20 committed:
Merge branch 'master' into vuluong/alert-ref

vuluongj20 committed:
fix(setup-docs): Use showIcon prop for Alert

vuluongj20 committed:
Merge branch 'master' into vuluong/alert-ref-trailing-items
